### ISSUE TITLE
instance.getConfig() should not allow mutations where it's not supported

### ISSUE THREAD

**Issue by jerrinot** [hazelcast/hazelcast#10812]
`instance.getConfig()` will give you a configuration object which happily allow mutations, but they are ignored. 

Hazelcast 3.9 allow to submit new configurations. All other mutation methods should throw UOE. 

Concern: Compatibility. For example WAN calls `addWanConfig()` on a config object while a member is already running.

**Comment by mmedenjak** [comment]
@jerrinot I can see some setters are already throwing exceptions. Are we talking about expanding this to all setters in the `DynamicConfigurationAwareConfig` or even to the other config classes returned by the getters?

**Comment by mdogan** [comment]
Ideally we should disallow all mutations in all config classes but it needs a massive work and testing. We already have readonly configs for some classes, we can use them. But for other classes, such as `GroupConfig`, we need to figure out things.

**Comment by mmedenjak** [comment]
@mdogan Right, I agree that it should be as complete as possible. Given that into consideration, I don't think this is necessary (or even fully possible) to do in 3.9 or even 3.x for that sake. I think @jerrinot has a point, because of compatibility we are very restricted. Even though it makes no sense, someone could have been mutating the config after the instance has been constructed and the new hazelcast version will start throwing `UnsupportedOperationException`.

**Comment by mdogan** [comment]
@mmedenjak & @jerrinot: are we going to move this out of 3.9?

**Comment by mmedenjak** [comment]
I'm for moving it. The choice of target version depends on whether we can introduce behaviour changes in a minor release, specifically throwing an UOE where we previously hadn't. If we can throw UOE from setters in a minor release, then 3.10, otherwise 4.0. 
OTOH we are already introducing this behaviour change in 3.9 as this has different behaviour in 3.8 and 3.9 (3.8 doesn't throw an exception) : 
`instance.getConfig().setExecutorConfigs(Collections.<String, ExecutorConfig>emptyMap())`

**Comment by jerrinot** [comment]
@mmedenjak, @mdogan: moved

**Comment by mmedenjak** [comment]
+1 for fixing in 4.0 as a part of config cleanup. Related: https://github.com/hazelcast/hazelcast/issues/12617

**Comment by puzpuzpuz** [comment]
One of our customers reported a related issue recently. See https://hazelcast.zendesk.com/agent/tickets/6221 for more details.
